### PR TITLE
Datapackage.json format

### PR DESCRIPTION
**Pull request by goelakash**
@henrykironde Refactored and added docstrings. I would prefer if I could add this PR to a separate dev branch, instead of `master`.

### CODE REVIEW

File: lib/parse_script_to_json.py [hunk 1]
```diff
@@ -1,114 +1,125 @@
...
+                if last_table:
+                    vs = [v.strip() for v in value.split(',')]
+
+                    if "fields" not in last_table["schema"]:
+                        last_table["schema"]["fields"] = []
+
+                    column = {}
+                    column['name'] = vs[0]
+                    column['type'] = vs[1]
+                    if len(vs)>2:
```
Comment (henrykironde) [inline lib/parse_script_to_json.py]: space needed if len(vs) > 2

**Comment by ethanwhite** [comment]
> I would prefer if I could add this PR to a separate dev branch, instead of master.

I just create a new `data-package` branch for you.

**Comment by ethanwhite** [comment]
There's one big overarching comment the could use some work before a fuller review, and that's that the new major functions are too long. This makes it difficult to follow everything that's happening while reading through the function which makes maintenance and debugging more difficult down the road. This is particularly true because of the number of layers of nesting of loops and conditionals in the functions.

In terms of rough rules of thumb, you want the largest functions to fit on a single screen (<40 lines) and ideally not have more than 3 layers of nesting if possible. To accomplish this take chunks of the current long functions and make them their own functions. Just for example, you could take the logic in the `t_key == "dialect"` code chunk and make it a function and the logic in the `elif t_key == "schema":` chunk and put it in a function. If these functions are well named then the function that calls them will read like a series of easy to understand commands and then the details are hidden in the functions that it calls.

If all that makes sense and you want to give it a go breaking these up, that's great. If you'd like me to go through one of the big functions in detail and talk more about exactly what I'd recommend I'd be happy to do that as well.

**Comment by ethanwhite** [comment]
Also, just wanted to say that I'm super excited to be getting close on this. Keep up the great work!

File: __main__.py [hunk 2]
```diff
@@ -78,6 +78,38 @@ def main():
...
             return
 
+        elif args.command == 'new_json':
+            # create new JSON script
+            create_json()
+            return
+
+        elif args.command == 'edit_json':
+            # edit existing JSON script
+            if args.dataset is None:
```
Comment (henrykironde) [inline __main__.py]: we can use `argparse` in [get_opts.py](https://github.com/weecology/retriever/blob/master/lib/get_opts.py) to make sure `edit_json` has a dataset argument.

File: lib/compile.py [hunk 3]
```diff
@@ -139,3 +141,125 @@ def get_value(key):
...
+                table_dict = {}
+
+                if table["schema"] == {} and table["dialect"] == {}:
+                    continue
+
+                for (t_key, t_val) in table.items():
+
+                    if t_key == "dialect":
+                        for (d_key, d_val) in t_val.items():
+                            #dialect related key-value pairs
```
Comment (henrykironde) [inline lib/compile.py]: space

File: lib/compile.py [hunk 4]
```diff
@@ -139,3 +141,125 @@ def get_value(key):
...
+                            if d_key == "nulls":
+                                table_dict['cleanup'] = "Cleanup(correct_invalid_value, nulls=" + str(d_val) + ")"
+
+                            elif d_key == "delimiter":
+                                table_dict[d_key] = "'"+str(d_val)+"'"
+                            else:
+                                table_dict[d_key] = d_val
+
+                    elif t_key == "schema":
+                        for (s_key, s_val) in t_val.items():
```
Comment (henrykironde) [inline lib/compile.py]: do we need the brackets (s_key, s_val) ?

File: lib/compile.py [hunk 5]
```diff
@@ -139,3 +141,125 @@ def get_value(key):
...
+    The scripts are created with `retriever create_json <script_name` using
+    command line
+    '''
+    json_object = yaml.safe_load(open(json_file + ".json","r"))
+
+    tables = {}
+    values = {}
+    keys_to_ignore = ["template"]
+
+    for (key,value) in json_object.items():
```
Comment (henrykironde) [inline lib/compile.py]: do we need the ()?

File: lib/compile.py [hunk 6]
```diff
@@ -139,3 +141,125 @@ def get_value(key):
...
+
+                    if t_key == "dialect":
+                        for (d_key, d_val) in t_val.items():
+                            #dialect related key-value pairs
+                            #copied as is
+                            if d_key == "nulls":
+                                table_dict['cleanup'] = "Cleanup(correct_invalid_value, nulls=" + str(d_val) + ")"
+
+                            elif d_key == "delimiter":
+                                table_dict[d_key] = "'"+str(d_val)+"'"
```
Comment (henrykironde) [inline lib/compile.py]: "'" + str(d_val) + "'"

File: lib/datapackage.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+short_names = [script.shortname.lower() for script in SCRIPT_LIST()]
+
+def create_json():
+    '''
+    Creates datapackage.JSON script by taking input from user via command line.
+    http://specs.frictionlessdata.io/data-packages/#descriptor-datapackagejson
+
+    Usage: retriever create_json <new_script_name>
+    '''
+    contents = {}
```
Comment (henrykironde) [inline lib/datapackage.py]:
```
'''
Creates datapackage JSON script

Take input from user via command line. 
http://specs.frictionlessdata.io/data-packages/#descriptor-datapackagejson
Usage: retriever create_json <new_script_name>
'''
```

File: lib/datapackage.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+                            if len(values) == 1:
+                                table['dialect'][key] = values[0]           # not a list type for a single value (bool,int, etc)
+                            else:
+                                table['dialect'][key] = values
+                            break
+
+                    elif i<6:
+                        while True:
+                            try:
+                                if val == "":       #User wants to skip
```
Comment (henrykironde) [inline lib/datapackage.py]: #User wants to skip--># User wants to skip

File: lib/compile.py (hunk 4 above)
Comment (goelakash) [inline lib/compile.py]: Makes the code readable.

File: lib/compile.py (hunk 5 above)
Comment (goelakash) [inline lib/compile.py]: Readability.

**Comment by goelakash** [comment]
@ethanwhite @henrykironde Thanks for the comments. I did a refactor of `compile_json` and some bug-fixing in `datapackage.py` module. Do we need tests for the CLI?

**Comment by henrykironde** [comment]
If there is a way to test CLI , then it will be nice to have the tests.
I don't know how we shall provide the input values but if there is away then it will be good.

**Comment by henrykironde** [comment]
If there is a way to test CLI , then it will be nice to have the tests.
I don't know how we shall provide the input values but if there is away then it will be good.

File: lib/datapackage.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+                except:
+                    print("Exception occured. Check the input format again.\n")
+                    pass
+
+                col = input()
+
+
+            isCT = input("Add crosstab columns? (y,N): ")
+            if isCT.lower() in ["y", "Y"]:
```
Comment (henrykironde) [inline lib/datapackage.py]: why are we checking for both lower and upper case?

File: lib/datapackage.py [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+        print("\nScript written to "+file_name)
+        output_file.close()
+
+
+def edit_dict(obj, tabwidth=0):
+    '''
+    Recursive helper function for edit_json() to edit a datapackage.JSON script file.
+    '''
+    for (key, val) in obj.items():
+        print('\n'+"  "*tabwidth+"->"+key+" : \n")
```
Comment (henrykironde) [inline lib/datapackage.py]: print('\n' + "  " *tabwidth + "->" + key + " : \n")

File: lib/datapackage.py [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+        if type(val) == list:
+            for v in val:
+                print("  "*tabwidth+str(v)+'\n')
```
Comment (henrykironde) [inline lib/datapackage.py]: operator spacing

File: lib/datapackage.py [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+        elif type(val) == dict:
+            for item in val.items():
+                print("  "*tabwidth+str(item)+'\n')
```
Comment (henrykironde) [inline lib/datapackage.py]: space before and after operator

File: lib/datapackage.py [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+        else:
+            print("  "*tabwidth+str(val)+'\n')
+
+        while True:
+            try:
+                if isinstance(val,dict):
+                    print("\n\t'"+key+"' has the following keys:\n"+str(obj[key].keys())+"\n")
```
Comment (henrykironde) [inline lib/datapackage.py]: for the spaces in between operators, if there is a reason for that then ignore them 👍

File: lib/datapackage.py [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+                print("  "*tabwidth+str(v)+'\n')
+        elif type(val) == dict:
+            for item in val.items():
+                print("  "*tabwidth+str(item)+'\n')
+        else:
+            print("  "*tabwidth+str(val)+'\n')
+
+        while True:
+            try:
+                if isinstance(val,dict):
```
Comment (henrykironde) [inline lib/datapackage.py]: space after ,

File: lib/datapackage.py [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+                    print("\n\t'"+key+"' has the following keys:\n"+str(obj[key].keys())+"\n")
+                    do_edit = input("Edit the values for these sub-keys of "+key+"? (y/N): ")
+
+                    if do_edit.lower() in ['y', 'yes']:
+                        edit_dict(obj[key], tabwidth+1)
```
Comment (henrykironde) [inline lib/datapackage.py]: space needed tabwidth + 1

File: lib/datapackage.py [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+
+                    print("Select one of the following for the key '"+key+"': \n")
```
Comment (henrykironde) [inline lib/datapackage.py]: space needed

File: lib/datapackage.py [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+
+                    elif selection == '3':
+                        del_key = input('Enter key to be deleted: ')
+                        if del_key not in val:
+                            print("Invalid key: Not found")
+                            continue
+                        print("Removed "+str(del_key) + " : " + str(obj[key].pop(del_key)))
+
+                    elif selection == '4':
+                        do_remove = input("Are you sure (Yes, completely remove this entry)? (y/n): ")
```
Comment (henrykironde) [inline lib/datapackage.py]: `input("Are you sure (Yes or Y,  will completely remove this entry)? (y/n): ")`

File: lib/datapackage.py [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+                            print("Removed "+key+" from script.\n")
+                        else:
+                            print("Aborted.")
+                            sleep(1)
+                    elif selection == '5' or selection == "":
+                        pass
+                    else:
+                        raise RuntimeError("Invalid input!")
+
+                elif isinstance(val,list):
```
Comment (henrykironde) [inline lib/datapackage.py]: `isinstance(val, list): `spacing needed

File: lib/datapackage.py [hunk 19]
```diff
@@ -0,0 +1,335 @@
...
+
+                    for i in range(len(val)):
+                        print(str(val[i]))
+                        if isinstance(val[i], dict):
+                            print("\n\t'"+key+"' has the following dict:\n"+str(val[i])+"\n")
```
Comment (henrykironde) [inline lib/datapackage.py]: make a check on all `prints` to see the spacing

File: lib/datapackage.py [hunk 20]
```diff
@@ -0,0 +1,335 @@
...
+                            do_edit = input("Edit the dict in '"+key+"'? (y/N): ")
```
Comment (henrykironde) [inline lib/datapackage.py]: spacing `'"+key+"`

File: lib/datapackage.py [hunk 21]
```diff
@@ -0,0 +1,335 @@
...
+
+                            if do_edit.lower() in ['y', 'yes']:
+                                edit_dict(obj[key][i], tabwidth+2)
```
Comment (henrykironde) [inline lib/datapackage.py]: spacing

File: lib/datapackage.py [hunk 22]
```diff
@@ -0,0 +1,335 @@
...
+
+                    elif selection == '2':
+                        del_val = input('Enter value to be deleted: ')
+                        if del_val not in obj[key]:
+                            print("Invalid value: Not found.")
+                            continue
+                        print("Removed "+str(obj[key].pop(del_key)))
+
+                    elif selection == '3':
+                        do_remove = input("Are you sure (Yes, completely remove this entry)? (y/n): ")
```
Comment (henrykironde) [inline lib/datapackage.py]: like in the line [228](https://github.com/weecology/retriever/pull/596/files#diff-fe19ce9787a421223628d4355dc2f5e4R228)
personally it reads better if we say 
`input("Are you sure (Yes/y will, completely remove this entry)? (y/n): ")`
but we could do as you wish

**Comment by henrykironde** [comment]
Note: I have not tested the whole program. I am still reading through

**Comment by henrykironde** [comment]
Additionally, in some areas we are using single char variable names, (v, i, k ...), this could be fine but becomes hard to follow especially for long code segments. Where possible we should try to give some unique names pertaining to what is being handled. Personally I do this all the time but I try to refactor when finalizing with that branch changes. Overall good so far.

File: lib/datapackage.py [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+
+    Takes input from user via command line.
+
+    Usage: retriever create_json
+    '''
+    contents = {}
+
+    script_exists = True
+    while script_exists:
+        contents['name'] = input("Shortname (unique identifier for script): ")
```
Comment (henrykironde) [inline lib/datapackage.py]:
```
input("Shortname (unique identifier for script): ")
input("Shortname (Give a unique identifier for script)
```

**Comment by henrykironde** [comment]
**test retriever new_json**

`retriever new_json`

```
Shortname (unique identifier for script): Test987
Title/Name: test script json
Description: Testing the json pack
Citation: at el a'Kash (Delhi)
Site/Homepage of dataset:
Tags (separated by commas):

Add Table? (y/N): testone
Not a valid option


Add Table? (y/N): y
Table name: testone
nulls (separated by ';'):
replace_columns (separated by ';'):
delimiter:
do_not_bulk_insert (bool = True/False):
contains_pk (bool = True/False):
escape_single_quotes (bool = True/False):
header_rows (int):
Enter columns [format = name, type, (optional) size]:

a,int,
a,char
b,char
c,char
d,char

Add crosstab columns? (y,N): n

Add Table? (y/N): n

Script written to Test987.json
```

we should prevent the creation of a script with  two columns having the same name in this case `a`



- It seems like the `url` is not accounted for in the resources.  

In some cases, when a user doesn't provide some information like tags, I think we should not create those items in the json file

**sample out put:**

```
{
    "citation": "at el a'Kash (Delhi)",
    "description": "Testing the json pack",
    "homepage": "",
    "keywords": "",
    "name": "Test987",
    "resources": [
        {
            "dialect": {},
            "name": "testone",
            "schema": {
                "fields": [
                    {
                        "name": "a",
                        "type": "int"
                    },
                    {
                        "name": "a",
                        "type": "char"
                    },
                    {
                        "name": "b",
                        "type": "char"
                    },
                    {
                        "name": "c",
                        "type": "char"
                    },
                    {
                        "name": "d",
                        "type": "char"
                    }
                ]
            }
        }
    ],
    "title": "test script json"
}
```

I am not sure what this get `header_rows (int):`

File: lib/datapackage.py (hunk 9 above)
Comment (goelakash) [inline lib/datapackage.py]: yeah, I should add a `strip()` after that.

**Comment by goelakash** [comment]
@henrykironde I added the dialect fields based on the members of `Table` class.

**Comment by henrykironde** [comment]
@goelakash. Let me go through the changes.  You know we have like 10 more days left so I am putting a hold on most of the work here so that we can get this done. Any help you need I will be available most of the time. :+1:  :+1:

**Comment by henrykironde** [comment]
can you give me your example of a sample run on Retriever new_json  and if possible make also one for edit Json

**Comment by goelakash** [comment]
```
Shortname (Give a unique identifier for script): new_test
Title/Name: New test json script

Description: balh

Citation: blah

Site/Homepage of dataset: ww.blah.com

Tags (separated by commas): ,;.;'\t'



Add Table? (y/N): y

Table name: table1

Table URL: www.table1.com

nulls (separated by ';'): , ; :

replace_columns (separated by ';'):

delimiter: \t

do_not_bulk_insert (bool = True/False): True

contains_pk (bool = True/False):

escape_single_quotes (bool = True/False):

escape_double_quotes (bool = True/False):

fixed_width (bool = True/False): False

header_rows (int):

Enter columns [format = name, type, (optional) size]:



c1, int, 12

x2, chae r, 1



Add crosstab columns? (y,N): n

Add Table? (y/N): n

Script written to new_test.json
```
Output:
```
{
    "citation": "blah",
    "description": "balh",
    "homepage": "ww.blah.com",
    "keywords": ",;.;'\\t'",
    "name": "new_test",
    "resources": [
        {
            "dialect": {
                "delimiter": "\\t",
                "do_not_bulk_insert": "True",
                "fixed_width": "False",
                "nulls": [
                    ",",
                    ":"
                ]
            },
            "name": "table1",
            "schema": {
                "fields": [
                    {
                        "name": "c1",
                        "size": "12",
                        "type": "int"
                    },
                    {
                        "name": "x2",
                        "size": "1",
                        "type": "char"
                    }
                ]
            },
            "url": "www.table1.com"
        }
    ],
    "title": "New test json script"
}
```

**Comment by henrykironde** [comment]
Okey we are creating the json scripts well so far, 

Can we go ahead and remove all the .script files so that we use only json

**Comment by henrykironde** [comment]
Note: you added  "url" and the json is created well but we dont have the same in the python script.
test below
```
{
    "citation": "",
    "description": "",
    "homepage": "",
    "keywords": "",
    "name": "newtest",
    "resources": [
        {
            "dialect": {},
            "name": "testone",
            "schema": {
                "fields": []
            },
            "url": "file:///C:/Users/Henry/Documents/GitHub/retriever/test.csv"
        }
    ],
    "title": "new test"
}
```

**newtest.py** 

```
#retriever
from retriever.lib.templates import BasicTextTemplate
from retriever.lib.models import Table, Cleanup, correct_invalid_value

SCRIPT = BasicTextTemplate(tables={'testone': Table('testone', columns=[])},
                           description="",
                           tags=,
                           citation="",
                           shortname="newtest",
                           ref="",
                           name="new test")
```

File: lib/compile.py [hunk 24]
```diff
@@ -139,3 +139,151 @@ def get_value(key):
...
+    '''
+    json_object = yaml.safe_load(open(json_file + ".json", "r"))
+
+    values = {}
+    values['urls'] = {}
+
+    keys_to_ignore = ["template"]
+
+    for (key, value) in json_object.items():
+
```
Comment (henrykironde) [inline lib/compile.py]: at this point we have a problem
```
> retriever ls
Traceback (most recent call last):
  File "C:\Python27\Scripts\retriever-script.py", line 9, in <module>
    load_entry_point('retriever===1.9.0-de', 'console_scripts', 'retriever')()
  File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 558, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 2682, in load_entry_point
    return ep.load()
  File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 2355, in load
    return self.resolve()
  File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 2361, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "C:\Python27\lib\site-packages\retriever-1.9.0_de-py2.7.egg\retriever\__main__.py", line 28, in <module>
    from retriever.lib.datapackage import create_json, edit_json
  File "C:\Python27\lib\site-packages\retriever-1.9.0_de-py2.7.egg\retriever\lib\datapackage.py", line 9, in <module>
    short_names = [script.shortname.lower() for script in SCRIPT_LIST()]
  File "C:\Python27\lib\site-packages\retriever-1.9.0_de-py2.7.egg\retriever\__init__.py", line 103, in SCRIPT_LIST
    return [module.SCRIPT for module in MODULE_LIST(force_compile)]
  File "C:\Python27\lib\site-packages\retriever-1.9.0_de-py2.7.egg\retriever\__init__.py", line 79, in MODULE_LIST
    compile_json(join(search_path, script_name))
  File "C:\Python27\lib\site-packages\retriever-1.9.0_de-py2.7.egg\retriever\lib\compile.py", line 211, in compile_json
    for (key, value) in json_object.items():
AttributeError: 'list' object has no attribute 'items'
C:\Users\Henry\Documents\GitHub\retriever [henry-datapack +22 ~0 -0 !]>
```

**Comment by goelakash** [comment]
@henrykironde can you do a clean install now and test? I think it was caused by wrong parsing of keywords/tags.

**Comment by henrykironde** [comment]
okey test and still the same thing happens` for (key, value) in json_object.items():`

**Comment by goelakash** [comment]
@henrykironde I am unable to reproduce the error. I carry out the following steps to do a fresh install on my machine:
1. Uninstall:
```
sudo python setup.py install --record files.txt
cat files.txt | sudo xargs rm -rf
```
2. Remove dirs
`sudo rm -r ~/.retriever`
3. Reinstall: 
```
git clone git@github.com:goelakash/retriever.git
cd retriever
sudo python setup.py install
```

**Comment by henrykironde** [comment]
I guess I have some scripts that I had created prior and they have errors. 
I will merge this

**Comment by henrykironde** [comment]
can you create the PR on the other branch that Ethan created other than master

**Comment by henrykironde** [comment]
@ethanwhite, said he created a branch `data-package`.

**Comment by goelakash** [comment]
Closing to create a new PR to datapackage branch
